File: android/app/src/main/java/com/example/ubapp/turnbased/Minimax.kt
package com.example.ubapp.turnbased

import kotlin.math.max
import kotlin.math.min

object Minimax {
    fun bestMove(model: TicTacToeModel, ai: Mark): Int? {
        var bestScore = Int.MIN_VALUE
        var bestIdx: Int? = null
        for (i in 0..8) {
            if (model.board[i] != 0) continue
            val copy = model.copy()
            copy.apply(i)
            val score = score(copy, ai, false, 0)
            if (score > bestScore) {
                bestScore = score
                bestIdx = i
            }
        }
        return bestIdx
    }

    private fun score(model: TicTacToeModel, ai: Mark, maximizing: Boolean, depth: Int): Int {
        model.winner?.let { return if (it == ai) 100 - depth else depth - 100 }
        if (model.isDraw) return 0

        var best = if (maximizing) Int.MIN_VALUE else Int.MAX_VALUE
        for (i in 0..8) {
            if (model.board[i] != 0) continue
            val copy = model.copy()
            copy.apply(i)
            val s = score(copy, ai, !maximizing, depth + 1)
            best = if (maximizing) max(best, s) else min(best, s)
        }
        return best
    }
}

File: lib/turnbased/minimax.dart
import 'tic_tac_toe_model.dart';

int? bestMove(TicTacToeModel model, Mark ai) {
  var bestScore = -1 << 30;
  int? bestIdx;
  for (var i = 0; i < 9; i++) {
    if (model.board[i] != Mark.empty) continue;
    final copy = model.copy();
    copy.apply(i);
    final s = _score(copy, ai, false, 0);
    if (s > bestScore) {
      bestScore = s;
      bestIdx = i;
    }
  }
  return bestIdx;
}

int _score(TicTacToeModel m, Mark ai, bool maximizing, int depth) {
  final w = m.winner;
  if (w != null) return w == ai ? 100 - depth : depth - 100;
  if (m.isDraw) return 0;

  var best = maximizing ? -1 << 30 : 1 << 30;
  for (var i = 0; i < 9; i++) {
    if (m.board[i] != Mark.empty) continue;
    final copy = m.copy();
    copy.apply(i);
    final s = _score(copy, ai, !maximizing, depth + 1);
    best = maximizing ? (s > best ? s : best) : (s < best ? s : best);
  }
  return best;
}
